Handle situations where the server misbehaves better

indianajohn opened the pull request:
### Contains

In #2230, @Cervator requested several things. This PR is meant to address that issue.

 - I tested #1626 in many different ways on the local host and was unable to reproduce the error. I'm not sure if it was fixed since, or if it only affects WAN downloads.

 - When I went into the relevant sources I found that some information about the module downloads (the version, name, and size) was being sent to the loading screen. Maybe somebody got to this before me?

 - I did, however, add a similar statement to the log.

 - I also added the count of how many modules remain to the output string that is sent to the loading screen while a module is being downloaded.

 - Finally, and most significantly, I added a time out to the server info & module download process. If no packets are received from the server for 10 seconds, the client gives up and returns to the server menu. From what I can gather, this situation never happens during the connection process unless something goes wrong; I was unable to make the time out occur unless I artificially induced a delay on the server side, so I don't think this PR adds any bugs.

 - This required a small adjustment to NetworkSystemImpl.java. Before my change, if JoinStatus.FAILED was returned by ClientConnectionHandler.getJoinStatus(), the client would be left in a state where it could not connect to any servers until the entire program was restarted. After this change, it's possible to connect to another server after a connection fails.

 - Finally, I added an error message if the server fails to transfer a needed module to the client. A dev should know if they make a change that causes that situation.

The net total of all of these changes is that, if #1626 does happen again, the user will not have to restart their client, and will also have more meaningful error messages.

### How to test

At the top of ```sendModules()``` in ```ServerConnectionHandler.java```, you can put the following code:

```java
logger.info("purposefully waiting to simulate a bug.");
try {
  Thread.sleep(20000);
} catch (InterruptedException e1) {
  e1.printStackTrace();
}
```

Doing this will cause the server to stall for 20 seconds while trying to send a module. Trying this before and after my change will lead to different outcomes.

 - Before, the loading screen will hang indefinitely.
 - After, the loading screen will exit after 10 seconds with a meaningful error message ('Server stopped responding") and a more technical error will be printed to the log ("Server timeout threshold of 10000 ms exceeded.").
 - There are also log messages posted for each download, and more detail on the loading screen.

### Outstanding before merging

- [ ] Need to decide whether a cancel button is still necessary. The cancel button seems like it will be a bigger change and might disturb other stuff that I don't fully understand (in the startup sequence), so I decided to leave it out of this PR.

GooeyHub commented:
Can one of the admins please verify this patch?

Cervator commented:
Excellent work! You are correct in suspecting some issues may get fixed over time even if by accident but not fully confirmed as fixed. That one I had expected as fixed but not taken the time to test exhaustively. Taking this opportunity to do so, although it got a bit lengthy! I could cut out a lot of the below, but I figured even after realizing I was doing something wrong still including the whole testing process might be helpful.

Testing on Win7 trying a headless server from old source with all stable modules, configured to run JoshariasSurvival with a client running Alpha 2 via binary with all extra modules removed. This intentionally causes version conflicts even on engine, Core, etc, since the source versions are +0.0.1-SNAPSHOT vs the binary.

With the debug snippet above added I managed to actually get disconnected while trying to download modules and returned to the menu. That then resulted in triggering #1775 - which I suspect this may also fix thanks to your networking tweak

Trying it again without the debug snippet hit the same logged error, but now the window stalled entirely like I've seen before, although I'm not sure I've noticed the log statement - different issue somehow? Log is from the server.

```
14:21:26.888 [New I/O worker #2] ERROR o.t.n.i.ServerConnectionHandler - Error sending module
java.nio.file.AccessDeniedException: J:\Dev\Terasology\Git\integrate\Terasology\modules\PotentialEnergyDevices
	at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:83)
	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:97)
	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:102)
	at sun.nio.fs.WindowsFileSystemProvider.newByteChannel(WindowsFileSystemProvider.java:230)
	at java.nio.file.Files.newByteChannel(Files.java:361)
	at java.nio.file.Files.newByteChannel(Files.java:407)
	at java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:384)
	at java.nio.file.Files.newInputStream(Files.java:152)
	at org.terasology.network.internal.ServerConnectionHandler.sendModules(ServerConnectionHandler.java:112)
	at org.terasology.network.internal.ServerConnectionHandler.messageReceived(ServerConnectionHandler.java:80)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
	at org.jboss.netty.handler.codec.oneone.OneToOneDecoder.handleUpstream(OneToOneDecoder.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:303)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.terasology.network.internal.MetricRecordingHandler.messageReceived(MetricRecordingHandler.java:45)
	at org.jboss.netty.channel.SimpleChannelHandler.handleUpstream(SimpleChannelHandler.java:88)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
	at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:337)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
	at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
	at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
	at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

Switching to the new code, same setup otherwise, I still hit the server error above then stall, which makes me think it is specific to my setup. The timeout doesn't trigger and I have to kill the client via task manager.

Worth mentioning is that some of these issues seem to happen with a server in a bad state. In one instance a server had seemingly run out of memory while in another an attempted reset hadn't gone right. We probably can't prevent all those issues but improving usability is a huge win that could be good enough to close the issues.

Next test with new code + debug snippet to try the stall. Same server side error, as expected I do get placed back in the main menu, however the network system remains active, doh, so still can't connect or dispose fully without task manager, boo

Realized this might simply be from not supporting module auto-download from source since there the module is a source directory not a jar file. Switched around to where the Alpha 2 binary is hosting the server and I'm running with the new code from source with the client. Got the regular stall in the client, waited a while, then killed the server via task manager. Client window then returned with an error. Still hit #1775. Drat.

Built new binary (`gradlew distApp`) to use as test server, swapping in the modules from Alpha 2. Still running from source with the client. Server started with minimal modules (running CoreSampleGameplay by default). Still getting intentional module version conflicts on Core and CoreSampleGameplay, but matching engine since I built that from source.

Worked! Despite actually leaving the 20 second module delay in place (and noticing it took that long before anything started happening). Huh. No timeout triggered?

Tried it again with a binary client with all modules removed, managed to download everything and get on the server, still with the 20 second delay in place. Odd, but probably enough to say that at least the module downloads work. Also got to see the new logging:

```
15:52:05.799 [New I/O worker #3] INFO  o.t.n.i.ClientHandshakeHandler - Received server verification
15:52:05.817 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Received server info
15:52:25.827 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading potentialenergydevices:1.0.0-SNAPSHOT (5.94 KB,23 modules remain)
15:52:25.835 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading ingamehelp:0.1.0-SNAPSHOT (28.89 KB,22 modules remain)
15:52:25.847 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading oregeneration:0.1.1-SNAPSHOT (19.04 KB,21 modules remain)
15:52:26.362 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading manuallabor:1.0.1-SNAPSHOT (169.33 KB,20 modules remain)
15:52:26.388 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading caves:0.2.1-SNAPSHOT (9.56 KB,19 modules remain)
15:52:26.395 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading durability:0.1.1-SNAPSHOT (10.30 KB,18 modules remain)
15:52:26.401 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading workstation:1.0.0-SNAPSHOT (77.23 KB,17 modules remain)
15:52:26.415 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading worldlytooltip:0.1.1-SNAPSHOT (11.35 KB,16 modules remain)
15:52:26.422 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading hunger:0.1.1-SNAPSHOT (20.50 KB,15 modules remain)
15:52:26.430 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading simplefarming:0.1.1-SNAPSHOT (31.60 KB,14 modules remain)
15:52:26.438 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading core:1.1.0 (1.74 MB,13 modules remain)
15:52:26.543 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading structuralresources:0.2.1-SNAPSHOT (39.11 KB,12 modules remain)
15:52:26.550 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading workstationingamehelp:1.0.0-SNAPSHOT (10.40 KB,11 modules remain)
15:52:26.556 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading joshariassurvival:1.0.0-SNAPSHOT (31.99 KB,10 modules remain)
15:52:26.564 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading multiblock:0.1.1-SNAPSHOT (55.72 KB,9 modules remain)
15:52:26.574 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading substancematters:1.0.0-SNAPSHOT (29.99 KB,8 modules remain)
15:52:26.581 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading irlcorp:1.1.1-SNAPSHOT (142.76 KB,7 modules remain)
15:52:26.595 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading fluid:0.1.1-SNAPSHOT (28.40 KB,6 modules remain)
15:52:26.604 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading itemrendering:0.1.2-SNAPSHOT (36.48 KB,5 modules remain)
15:52:26.611 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading manuallaboreventualskills:1.0.0-SNAPSHOT (13.38 KB,4 modules remain)
15:52:26.616 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading machines:1.1.0-SNAPSHOT (130.00 KB,3 modules remain)
15:52:26.628 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading plantpack:0.1.1-SNAPSHOT (232.94 KB,2 modules remain)
15:52:26.647 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading eventualskills:1.0.0-SNAPSHOT (53.74 KB,1 modules remain)
15:52:26.656 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Downloading customoregen:0.1.1-SNAPSHOT (40.35 KB,0 modules remain)
15:52:26.672 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Join complete received
15:52:26.673 [New I/O worker #3] INFO  o.t.n.internal.NetworkSystemImpl - Connected to server
```

One interesting caveat I'll create a separate issue for is that modules downloaded this way will not have contributed any hot keys to the client on initialization. So since I had deleted all modules I couldn't use the inventory or JoshariasSurvivals help or skill systems. That's unrelated but the first time I've encountered it. Simply restarting then would leave everything well.

Assuming we can confirm or fix the timeout thing I think we can worry about a cancel button later, good call leaving it out :-)

Protip: If you use topic branches (preferred) or Git rebase you can avoid including merge commits.

indianajohn commented:
Hello again,

Terribly sorry for wasting your time. I have this funny habit of fixing something, testing it, deciding I'm done, then making a last-minute change before I commit it. In this case it broke everything and I was able to confirm that it didn't even work correctly on my machine in its state as of last night.

I have

 - Fixed it properly this time
 - squashed my commits together, removed the merge commit
 - Tested that the following sequence is possible on my machine now:
    - insert debug code on server,
    - build the server,
    - run the server,
    - run the client from source with an unclean module configuration,
    - see timeout message after 10 seconds, and
    - attempt to connect to the sever again.

I've also tested
 - that client connections work in general (that is, that timeouts do not occur when they're not desired), and
 - that it is possible to actually connect to the server after one connection attempt has failed, without restarting the client.

The issue of the program cleanly exiting with loose network threads hanging around is still outstanding, but not part of this issue/pull request IMO. I'll look at it later when I get some time ( #1775 ). It will require a bit more research as I think the problem is a bit further below the surface.

Cervator commented:
No worries on time, I really had to get back on top of how the related code worked anyway, this was a good excuse! :-)

Appreciate the fixes, will check soon. Would indeed be great to have the disposal issue checked on too and yeah separate PR is good there.

@GooeyHub: add to whitelist

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

Cervator commented:
Retested! Looking great :-)

One functional quirk noted: the timer thread on a failure spazzes out after an additional 20 seconds and prints a pile of log statements. The initial logging seems fine, but then the timer doesn't seem to exit right. Example:

```
22:26:57.772 [main] WARN  o.t.engine.internal.TimeBase - Delta too great (14955), capping to 1000
22:27:10.059 [ServerList Downloader] INFO  o.t.r.n.l.m.ServerListDownloader - Retrieved game server ServerInfo [name=mt.nerdpol.ovh, address=mt.nerdpol.ovh, port=25777, owner=Calinou, active=true]
22:27:10.309 [ServerList Downloader] INFO  o.t.r.n.l.m.ServerListDownloader - Retrieved game server ServerInfo [name=evorigin.tk, address=evorigin.tk, port=25777, owner=war-sloop, active=true]
22:27:10.559 [ServerList Downloader] INFO  o.t.r.n.l.m.ServerListDownloader - Retrieved game server ServerInfo [name=tuffkidtek, address=terasology.tuffkidtek.com, port=25777, owner=minnesotags, active=true]
22:27:12.641 [New I/O worker #1] INFO  o.t.n.i.ClientHandshakeHandler - Received Server Hello
22:27:12.646 [New I/O worker #1] INFO  o.t.n.i.ClientHandshakeHandler - Sending client certificate
22:27:12.842 [New I/O worker #1] INFO  o.t.n.i.ClientHandshakeHandler - Received server verification
22:27:12.872 [New I/O worker #1] INFO  o.t.n.i.ServerInfoRequestHandler - Received server info
22:27:14.382 [Parallel Operation] INFO  o.t.n.i.ClientConnectionHandler - constructed CCH
22:27:14.388 [New I/O worker #3] INFO  o.t.n.i.ClientHandshakeHandler - Received Server Hello
22:27:14.389 [New I/O worker #3] INFO  o.t.n.i.ClientHandshakeHandler - Sending client certificate
22:27:14.464 [New I/O worker #3] INFO  o.t.n.i.ClientHandshakeHandler - Received server verification
22:27:14.486 [New I/O worker #3] INFO  o.t.n.i.ClientConnectionHandler - Received server info
22:27:14.496 [main] ERROR org.terasology.assets.AssetType - Asset instances remained loaded after disposal - [engine:waitPopup!instance]
22:27:24.582 [Timer-0] INFO  o.t.n.i.ClientConnectionHandler - callback
22:27:24.582 [Timer-0] ERROR o.t.n.i.ClientConnectionHandler - Server timeout threshold of 10000 ms exceeded.
22:27:24.686 [Timer-0] ERROR o.t.n.i.ClientConnectionHandler - Server timeout threshold of 10000 ms exceeded.
22:27:44.712 [Timer-0] ERROR o.t.n.i.ClientConnectionHandler - Server timeout threshold of 10000 ms exceeded.
22:27:44.716 [Timer-0] ERROR o.t.n.i.ClientConnectionHandler - Server timeout threshold of 10000 ms exceeded.
22:27:44.717 [Timer-0] ERROR o.t.n.i.ClientConnectionHandler - Server timeout threshold of 10000 ms exceeded.
...
```

Note the callback + two log statements followed by the 20 second pause, then hundreds of timeout messages follow in rapid succession. That's weird!

Only other thing noted: whitespace is a bit off and your editor is set to tabs instead of spaces. Code convention is 4 space indents. Mind fixing? :-)

I ran a quick Reformat Code in IntelliJ (CTRL-ALT-L) to try and that fixed it up, with a bit of oddities left in from a few unnecessary line breaks. One block ended up looking like this after me tweaking a bit:

```java
    public ClientConnectionHandler(JoinStatusImpl joinStatus, NetworkSystemImpl networkSystem) {
        this.networkSystem = networkSystem;
        this.joinStatus = joinStatus;
        this.moduleManager = CoreRegistry.get(ModuleManager.class);
        logger.info("constructed CCH");
        timeoutPoint = System.currentTimeMillis() + timeoutThreshold;
        timeoutTimer.schedule(new java.util.TimerTask() {
            @Override
            public void run() {
                synchronized (joinStatus) {
                    logger.info("callback");
                    if (System.currentTimeMillis() > timeoutPoint && joinStatus.getStatus() != JoinStatus.Status.COMPLETE) {
                        joinStatus.setErrorMessage("Server stopped responding.");
                        logger.error("Server timeout threshold of {} ms exceeded.", timeoutThreshold);
                    }
                }
            }
        }, timeoutThreshold + 200);
    }
```

Happy to merge after that. Thank you so much!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

indianajohn commented:
Hey,

The multiple message thing was easy to fix. I also knocked out the hard tabs and weird formatting. Sorry about that, I'm using Eclipse for the first time because my student license to idea expired a while ago.

I have also noticed the program hanging at the end. I'm not 100% sure how to fix it, but I will investigate. It's up to you whether you want to keep this PR open until I find the cause.

Thanks for your feedback/testing.

Cervator commented:
Thanks :-) No problem keeping this open for a bit.

IntelliJ Community Edition is free and still great. We have open source access to Ultimate but it doesn't really add anything needed.

indianajohn commented:
Okay, the most recent commit seems to resolve the issue where the server hangs upon exit after a failed network connection. Let me know if you have any problems with it.

Sorry for the delay, I got busy at work.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

Cervator commented:
Tested it out well! Network system closes nicely :-)

No problem at all on the timing. Work happens! So long that you eventually make it back we'll be happy :D 

Tomorrow we have a multiplayer test event (starting in roughly 10-11hours) so that'll be great timing to test this some more. Testing Alpha 2 some as well as some new stuff by @xrtariq2594 so I'll prep a build with something from him that then includes this and a custom built module with a new version on it.

Thanks a lot!

By the way, are you checking out the rendering related PRs now and then? They're dark magic to me but may be relevant to your future interests :-)

Cervator commented:
Doh I missed a spot. The extra shutdown cleanup that was previously skipped caused an NPE when you were returning from a *single player game* - I added a small tweak that now seems to cover both cases well, see the linked commit for details :-)

Cervator commented:
Also, @indianajohn, I *do* think this in the end fixed the connection issues keeping the game from disposing. It makes sense when looking at the changes we made. So thanks a bunch again for helping figure that out :-) Closed that issue too for now.
